src/code_gen.rs

```diff
@@ -12,25 +16,93 @@ impl<'a> CodeGen<'a> {
     }
 }
 
+fn find_node_helper<'a>(node: &'a mut TreeNode, id: &'a str) -> Option<&'a mut TreeNode> {
+    if node.get_id() == id {
+        return Some(node);
+    }
+    for child in node.mut_children().iter_mut() {
+        if let Some(node) = find_node_helper(child, id) {
+            return Some(node);
+        }
+    }
+
+    None
+}
+
+fn find_node_with_id<'a>(
+    nodes: &'a mut HashMap<&str, TreeNode>,
+    id: &'a str,
+) -> Option<&'a mut TreeNode> {
+    let node_opt = None;
+    for (_, v) in nodes.iter_mut() {
+        let res = find_node_helper(v, id);
+        if res.is_some() {
+            return res;
+        }
+    }
+
+    node_opt
+}
+
 impl<'a> TreeFold<'a> for CodeGen<'a> {
+    fn visit_patterns(&mut self, patterns: &'a Patterns) {
+        for pattern in &patterns.0 {
+            self.visit_pattern(pattern);
+        }
+        assert!(
+            self.nodes.len() == 1,
+            "Only supports tree pattern, got a forrest with {} roots",
+            self.nodes.len()
+        );
+    }
+
     fn visit_pattern(&mut self, pattern: &'a Pattern) {
-        let from_node = pattern.from_node.id_name;
-        let to_node = pattern.to_node.id_name;
-        if self.paths.is_empty() {
-            self.paths.push(vec![from_node, to_node]);
-        } else {
-            let mut inserted = false;
-            for path in &mut self.paths {
-                if path[0] == to_node {
-                    path.insert(0, from_node);
-                    inserted = true;
-                } else if path[path.len() - 1] == from_node {
-                    path.push(to_node);
-                    inserted = true;
-                }
+        // TODO: support PATH
+        let rel_type = &pattern.relationship_type;
+        if let Relationship::Path(_) = rel_type {
+            panic!("Path relationship type is not yet supported.");
+        }
+
+        let to_id = pattern.to_node.id_name;
+        let to_node = match self.nodes.remove(to_id) {
+            None => TreeNode {
+                id: to_id.to_string(),
+                ..Default::default()
+            },
+            Some(n) => n,
+        };
+
+        let from_id = pattern.from_node.id_name;
+        let from_node = match find_node_with_id(&mut self.nodes, from_id) {
+            Some(n) => n,
+            None => self.nodes.entry(from_id).or_insert_with(|| TreeNode {
+                id: from_id.to_string(),
+                ..Default::default()
+            }),
+        };
+
+        from_node.mut_children().push(to_node);
+    }
+
+    fn visit_filter(&mut self, filter: &'a Filter) {
```

Comment:
Add a comment on top of visit_pattern and visit_filter to clarify what its purpose is: to generate a TreeNode proto corresponding to the MATCH and WHERE clauses respectively.